scripts/vim-patch.sh

```diff
@@ -7,114 +7,192 @@ NEOVIM_SOURCE_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+  git commit --allow-empty --file - <<< "${neovim_message}" | xargs echo -n
+  echo -e "\r✔ " # Replace ✘ with ✔
+
+  echo
+  echo "Creating files."
+  echo "${vim_diff}" > ${NEOVIM_SOURCE_DIR}/${neovim_branch}.diff
+  echo "✔ Saved patch to '${NEOVIM_SOURCE_DIR}/${neovim_branch}.diff'."
+  echo "${vim_full}" > ${NEOVIM_SOURCE_DIR}/${neovim_branch}.patch
+  echo "✔ Saved full commit details to '${NEOVIM_SOURCE_DIR}/${neovim_branch}.patch'."
+  echo "${neovim_pr}" > ${NEOVIM_SOURCE_DIR}/${neovim_branch}.pr
+  echo "✔ Saved suggested PR description to '${NEOVIM_SOURCE_DIR}/${neovim_branch}.pr'."
+  echo "You can use 'git clean' to remove these files when you're done."
+
+  echo
+  echo "Instructions:"
+  echo
+  echo "  Proceed to port the patch."
+  echo "  You might want to try 'patch -p1 < ${neovim_branch}.diff' first."
+  echo
+  echo "  Stage your changes ('git add ...') and use 'git commit --amend' to commit."
+  echo
+  echo "  Push your changes with 'git push origin ${neovim_branch}' and create a"
+  echo "  pull request called '[RFC] vim-patch:${vim_version}'. You might want "
+  echo "  to use the text in '${neovim_branch}.pr' as the description of this pull request."
+  echo
+  echo "  See https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-vim"
+  echo "  for more information."
+}
+
+list_vim_patches() {
+  echo
+  echo "Vim patches missing from Neovim:"
+
+  # Get vim patches and runtime file updates.
+  # Start from 7.4.442. The runtime was re-integrated from 7.4.384, but
+  # runtime patches before between 384 and 442 have already been ported
+  # to Neovim as of the creation of this script.
+  local vim_commits=$(cd ${VIM_SOURCE_DIR} && \
+    hg log --removed --template='{if(startswith("Added tag", firstline(desc)),
+      "{latesttag}\n",
+      "{if(startswith(\"updated for version\", firstline(desc)),
+        \"\",
+        \"{node}\n\")}")}' -r tip:v7-4-442)
+  # Append remaining vim patches.
+  # Start from 7.4.160, where Neovim was forked.
+  local vim_old_commits=$(cd ${VIM_SOURCE_DIR} && \
+    hg log --removed --template='{if(startswith("Added tag",
+      firstline(desc)),
+      "{latesttag}\n")}' -r v7-4-442:v7-4-161)
+
+  local vim_commit
+  for vim_commit in ${vim_commits} ${vim_old_commits}; do
+    local is_missing
+    if [[ "${vim_commit}" =~ v([0-9]-[0-9]-([0-9]{3,4})) ]]; then
+      local patch_number="${BASH_REMATCH[2]}"
+      # "Proper" Vim patch
+      # Check version.c:
+      is_missing="$(sed -n '/static int included_patches/,/}/p' ${NEOVIM_SOURCE_DIR}/src/nvim/version.c |
+        grep -x -e "[[:space:]]*//${patch_number} NA" -e "[[:space:]]*${patch_number}," >/dev/null && echo "false" || echo "true")"
+      vim_commit="${BASH_REMATCH[1]//-/.}"
+    else
+      # Untagged Vim patch, e.g. runtime updates.
+      # Check Neovim log:
+      is_missing="$(cd ${NEOVIM_SOURCE_DIR} &&
+        git log --no-merges --grep="^vim\-patch:${vim_commit:0:7}" --pretty=format:"false")"
```

Comment:
Maybe the `^` should be removed from `--grep="^vim\-patch:${vim_commit:0:7}"` since "vim-patch" is not always the first thing in the commit message.